share/eds/eds.go

```diff
@@ -0,0 +1,260 @@
+	size := eds.Width() * eds.Width()
+	shares := make([][]byte, size)
+
+	quadrantWidth := int(eds.Width() / 2)
+	quadrantSize := quadrantWidth * quadrantWidth
+	for i := 0; i < quadrantWidth; i++ {
+		for j := 0; j < quadrantWidth; j++ {
+			cells := getQuadrantCells(eds, uint(i), uint(j))
+			innerOffset := i*quadrantWidth + j
+			for quadrant := 0; quadrant < 4; quadrant++ {
+				shares[(quadrant*quadrantSize)+innerOffset] = prependNamespace(quadrant, cells[quadrant])
+			}
+		}
+	}
+	return shares
+}
+
+// getQuadrantCells returns the cell of each EDS quadrant with the passed inner-quadrant coordinates
+func getQuadrantCells(eds *rsmt2d.ExtendedDataSquare, i, j uint) [][]byte {
+	cells := make([][]byte, 4)
+	quadrantWidth := eds.Width() / 2
+	cells[0] = eds.GetCell(i, j)
+	cells[1] = eds.GetCell(i, j+quadrantWidth)
+	cells[2] = eds.GetCell(i+quadrantWidth, j)
+	cells[3] = eds.GetCell(i+quadrantWidth, j+quadrantWidth)
+	return cells
+}
+
+// prependNamespace adds the namespace to the passed share if in the first quadrant,
+// otherwise it adds the ParitySharesNamespace to the beginning.
+func prependNamespace(quadrant int, share []byte) []byte {
+	switch quadrant {
+	case 0:
+		return append(share[:appconsts.NamespaceSize], share...)
+	case 1, 2, 3:
+		return append(appconsts.ParitySharesNamespaceID, share...)
+	default:
+		panic("invalid quadrant")
+	}
+}
+
+// rootsToCids converts the EDS's Row and Column roots to CIDs.
+func rootsToCids(eds *rsmt2d.ExtendedDataSquare) ([]cid.Cid, error) {
+	var err error
+	roots := append(eds.RowRoots(), eds.ColRoots()...)
+	rootCids := make([]cid.Cid, len(roots))
+	for i, r := range roots {
+		rootCids[i], err = ipld.CidFromNamespacedSha256(r)
+		if err != nil {
+			return nil, fmt.Errorf("failure to get cid from root: %w", err)
+		}
+	}
+	return rootCids, nil
+}
+
+// ReadEDS reads the first EDS quadrant (1/4) from an io.Reader CAR file.
+// Only the first quadrant will be read, which represents the original data.
+// The returned EDS is guaranteed to be full and valid against the DataRoot, otherwise ReadEDS errors.
+func ReadEDS(ctx context.Context, r io.Reader, root share.Root) (*rsmt2d.ExtendedDataSquare, error) {
+	carReader, err := car.NewCarReader(r)
+	if err != nil {
+		return nil, fmt.Errorf("share: failure to read car file: %w", err)
+	}
+
+	odsWidth := len(root.RowsRoots) / 2
+	odsSquareSize := odsWidth * odsWidth
+	shares := make([][]byte, odsSquareSize)
+	// the first quadrant is stored directly after the header,
+	// so we can just read the first odsSquareSize blocks
+	for i := 0; i < odsSquareSize; i++ {
+		block, err := carReader.Next()
+		if err != nil {
+			return nil, fmt.Errorf("share: failure to read next car entry: %w", err)
+		}
+		// the stored first quadrant shares are wrapped with the namespace twice.
+		// we cut it off here, because it is added again while importing to the tree below
+		shares[i] = block.RawData()[ipld.NamespaceSize:]
```

Comment:
nit: consistency across which namespace size constant is used

Reply:
done